### ISSUE TITLE
Take full-width punctuation marks as boundary for hyperlinks

### ISSUE THREAD

**Issue by petercpg** [mattermost/mattermost#13971]
#### Summary
For some languages, especially CJK, full-width punctuation marks are frequently used in messages mixed with half-width marks when the message contains Western languages. However only half-width symbol are detected as ending boundary correctly to generate hyperlink for URLs, When the link ends with a full-width symbol it was not considered as boundary, making the following characters included into the generated hyperlink.

This seems to be a Markdown parser issue that happens to my Mattermost Web (Team Edition v5.19.1), App (Windows v4.4.0), Mobile (Android v1.28.0 build 268).

#### Steps to reproduce
Sending following message in any channel
```
URL Test Half-width (http://google.com/)
URL Test Full-width（http://google.com/）
"http://google.com/"
「http://google.com/」
URL Test Long Sentence http://google.com/, https://mattermost.com/，這是第二個網址。
```

#### Expected behavior
All hyperlinks shall be generated without trailing %-encoded strings

#### Observed behavior (that appears unintentional) 
The hyperlinks are generated incorrectly when the URLs are ended with full-width punctuation marks. : 
![Screenshot](https://user-images.githubusercontent.com/205738/75656025-09f71f80-5c9e-11ea-81eb-948a479d7fab.png)

#### Possible fixes
If you can, link to the line of code that might be responsible for the problem

**Comment by amyblais** [comment]
@hmhealey Do you know if there is a bug here?

**Comment by hmhealey** [comment]
Yeah, this does appear to be a bug. GitHub even has the same issues with that text, although that makes sense since we've borrowed parts of this code from them in the past.

@petercpg Do you have a list of characters that you've seen this problem with? In your example above, I see `（`, `）`, `「`, `」`, and `，`.

**Comment by petercpg** [comment]
Few more I can think of: `。？！『』【】《》〈〉–…：；〝〞～`. Note these are only characters daily used in my locale (zh-TW), there are more used in others regardless of the language itself.

**Comment by petercpg** [comment]
Just came to my mind that [Unicode range U+3000–303F: CJK Symbols and Punctuation](https://www.unicode.org/charts/PDF/U3000.pdf) should help - while I think some marks like`？`,`！` were considered as variation of their half-width relatives so were listed other ranges.

**Comment by hmhealey** [comment]
Thanks, we'll have to take a look at that. I've filed a Jira ticket for us to follow up on it and prioritize accordingly.

https://mattermost.atlassian.net/browse/MM-23180

**Comment by willypuzzle** [comment]
Can I grab this?

**Comment by willypuzzle** [comment]
After a bit of investigation I got that the problem is in the [marked](https://github.com/chjj/marked) package that parses links in the wrong way. We should open a bug there.

@hanzei

**Comment by hmhealey** [comment]
@willypuzzle We've actually got our own fork of that package that's diverged pretty massively from the upstream one (it's a long story, but they didn't make changes for a while so I modified it heavily enough that we couldn't update later on). If you could open a PR against that repo, that'd be great!

Once that's open, you'd need to submit a second PR against this repo to update the git commit of the `marked` package to point to the commit with your changes so that we can use it here.

Our fork is at:
https://github.com/mattermost/marked

**Comment by willypuzzle** [comment]
> [@willypuzzle](https://github.com/willypuzzle) We've actually got our own fork of that package that's diverged pretty massively from the upstream one (it's a long story, but they didn't make changes for a while so I modified it heavily enough that we couldn't update later on). If you could open a PR against that repo, that'd be great!
> 
> Once that's open, you'd need to submit a second PR against this repo to update the git commit of the `marked` package to point to the commit with your changes so that we can use it here.
> 
> Our fork is at: https://github.com/mattermost/marked

@hmhealey ok! Thanks for the information! I'll take action soon!
